src/Firestore/DocumentSnapshot.php

```diff
@@ -0,0 +1,314 @@
+<?php
+/**
+ * Copyright 2017 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Google\Cloud\Firestore;
+
+/**
+ * Represents the data of a document at the time of retrieval.
+ * A snapshot is immutable and may point to a non-existing document.
+ *
+ * Fields may be read in array-style syntax. Note that writing using array-style
+ * syntax is NOT supported and will result in a `\BadMethodCallException`.
+ *
+ * Example:
+ * ```
+ * use Google\Cloud\Firestore\FirestoreClient;
+ *
+ * $firestore = new FirestoreClient();
+ * $document = $firestore->document('users/john');
+ * $snapshot = $document->snapshot();
+ * ```
+ *
+ * ```
+ * // Fields are exposed via array-style accessors:
+ * $bitcoinWalletValue = $snapshot['wallet']['cryptoCurrency']['bitcoin'];
+ * ```
+ */
+class DocumentSnapshot implements \ArrayAccess
+{
+    /**
+     * @var DocumentReference
+     */
+    private $reference;
+
+    /**
+     * @var ValueMapper
+     */
+    private $valueMapper;
+
+    /**
+     * @var array
+     */
+    private $info;
+
+    /**
+     * @var array
+     */
+    private $fields;
+
+    /**
+     * @var bool
+     */
+    private $exists;
+
+    /**
+     * @param DocumentReference $reference The document which created the snapshot.
+     * @param ValueMapper $valueMapper A Firestore Value Mapper.
+     * @param array $info Document information, such as create and update timestamps.
+     * @param array $fields Document field data.
+     * @param bool $exists Whether the document exists in the Firestore database.
+     */
+    public function __construct(
+        DocumentReference $reference,
+        ValueMapper $valueMapper,
+        array $info,
+        array $fields,
+        $exists
+    ) {
+        $this->reference = $reference;
+        $this->valueMapper = $valueMapper;
+        $this->info = $info;
+        $this->fields = $fields;
+        $this->exists = $exists;
+    }
+
+    /**
+     * Get the reference of the document which created the snapshot.
+     *
+     * Example:
+     * ```
+     * $reference = $snapshot->reference();
+     * ```
+     *
+     * @return DocumentReference
+     */
+    public function reference()
+    {
+        return $this->reference;
+    }
+
+    /**
+     * Get the document path.
+     *
+     * Example:
+     * ```
+     * $name = $snapshot->name();
+     * ```
+     *
+     * @return string
+     */
+    public function name()
```

Comment:
This is fixed now. I've added `path()` methods to `CollectionReference`, `DocumentReference` and `DocumentSnapshot`. We retain `name()` as well because we provide that method throughout all of google-cloud-php to expose full resource names. I added additional documentation to all three methods on all classes and fixed existing ambiguities in the docs.